### ISSUE TITLE
Convert all v1 teams to v2

### ISSUE THREAD

**Issue by bcmdarroch** [chef/automate#1144]
MIGHT NOT BE MIGRATING TEAMS WITH INVALID V2 IDS

https://chefio.slack.com/archives/C61F9HHKK/p1584134040307400

## Problem
Currently, the Teams database structure reflects the V1 Team definition, which featured an auto-generated UUID, a unique Name, and a Description that can be pretty much any combination of characters.
```
chef_teams_service=# select * from teams;
                  id                  |       name        |            description             |          created_at           |          updated_at           | deletable | projects | db_id 
--------------------------------------+-------------------+------------------------------------+-------------------------------+-------------------------------+-----------+----------+-------
 92f654c4-0160-47ba-bcfb-81edbdfd2363 | admins            | admins                             | 2020-01-13 20:15:55.681153+00 | 2020-01-13 20:17:46.694595+00 | f         | {}       |     1
 21e978ba-796f-454b-a991-64b5697174ed | viewers           | viewers                            | 2020-01-13 20:17:46.903343+00 | 2020-01-13 20:17:46.913058+00 | t         | {}       |     2
 d5055874-0438-40e9-8b0b-5a341d5d23d1 | editors           | editors                            | 2020-01-13 20:17:47.007353+00 | 2020-01-13 20:17:47.015082+00 | t         | {}       |     3
 315667d4-ddf1-4e15-aa84-3f835ee26340 | unique team name! | this is my unique team description | 2020-01-13 20:50:31.767362+00 | 2020-01-13 20:50:31.767362+00 | t         | {}       |     4
(4 rows)
```

When we introduced V2, we chose not to migrate V1 teams to a new V2 team table. Instead, the V2 API returns the V1 team properties under different field names.
```
V1 team
{
    "id": "315667d4-ddf1-4e15-aa84-3f835ee26340",
    "name": "unique team name!",
    "description": "this is my unique team description"
}

same team on V2
{
    "id": "unique team name!",
    "name": "this is my unique team description",
    "projects": []
}
```

Problems arise when the V1 name does not comply with V2 ID alphanumeric validation. Right now we do nothing to rectify this issue.
```
curl -k -H "api-token: $TOK" https://a2-dev.test/apis/iam/v2/teams/unique team name! | jq 
<curl errors>
{
  "error": "no team found with id \"unique\"",
  "code": 5,
  "message": "no team found with id \"unique\"",
  "details": []
}
```
Once we are ready to force-upgrade all customers to V2, we need to perform some database migrations on teams to:
- ensure the table structure reflects the V2 Team definition
- ensure all properties are valid under V2 validation rules

Related:

Team names that do not conform to validation also cause problems in the UI, resulting in a broken url that eventually freezes the page:

![Screen Shot 2020-02-13 at 1.59.12 PM.png](https://images.zenhubusercontent.com/5d24fdf9dc67db59a676ebc6/0dbdcccc-2986-41f2-8a64-e7e4727acfaa)

Or results in a broken page view since we can see the team on the list page, but not its details:

![Screen Shot 2020-02-13 at 2.00.37 PM.png](https://images.zenhubusercontent.com/5d24fdf9dc67db59a676ebc6/d3aa6f84-9f23-4e77-8ecd-773987dc7c09)

## Technical Details

@tylercloke: When the user opts into the beta by upgrading, we should run a migration on the name fields in the database. We will be using the name field as the new ID. Since we are using them as the ID, they should conform to the IDs of the other fields. 

Let's lower-case letters, drop any non-alphanumeric field, convert spaces to "-", and resolve any conflicts by adding a number at the end. For example, if I had "My Super Team @" and "My Super Team $", those would be migrated to "my-super-team" and "my-super-team1". 

## Definition of Done
- [ ] When upgraded to IAM v2, all v1 teams undergo the following permanent migration:
        - drop team column`name` -> `id`
        - convert all v1 team names to valid v2 id format (see above)
        - then rename team column`name` -> `id`
        - rename team column `description` -> `name`
- [ ] v1 teams that have the same name with different letter casings (i.e. `admins, Admins, aDmIns`) are migrated as separate teams with distinct v2 valid IDs (since ID format will force these names to be lower-cased and therefore lose uniqueness)
- [ ] Any policies that have a v1 team member should be updated with the team's new v2 ID. This will likely need to be done via cereal so that it's transactional across services.

## Demo Script / Repro Steps
- I upgrade to the latest version of Automate which includes the GA release of IAM v2
- As part of the successful upgrade, any teams I have created are converted to their new IAM v2 forms.

## Notes
- 08/06/2019 @blakestier: "if we eventually force everyone up to v2, this card will be much easier (downgrading complicates it a bit)"

**Comment by blakestier** [comment]
Tinkering with this a little bit:

1) The casing is migrated so we don't get a uniqueness violation (so if we make a `TeamName` and a `teamname`, they migrate successfully and can still be edited
2) The above is still a little weird, as we don't allow that kind of casing for _new_ teams in v2
3) We also migrate team names that contain spaces or other odd characters over (`Name Space` becomes the id `Name Space`). This creates some weirdness in the UI, but still allows updates.
4) We don't have the same validation constraints in the DB as we do in the protos, I believe. That's why migrating goes happily, but then again, that's why data in the v2 form is not in sync with our v2 standards.

**Comment by blakestier** [comment]
Screenshot with the UI weirdness for names with spaces: (I see this if I go to the team details view, click `Details` then click _back_ to `Users`
<img width="1684" alt="Screen Shot 2019-10-08 at 10 56 02 AM" src="https://user-images.githubusercontent.com/24259050/66420263-71909380-e9ba-11e9-9611-6ca30b5392de.png">

**Comment by blakestier** [comment]
Should we consider enforcing our proto validations at the database level as well?

**Comment by susanev** [comment]
from @tylercloke 
we can no longer repro this issue.

the UI previously failed to encode special URL characters like spaces correctly. this is no longer the case. teams with special characters and spaces are now properly encoded in both the UI and the API requests the UI made. for example, making a PUT request to a team with spaces for the ID resulted in this request being sent from the UI via this URL:
https://a2-dev.test/apis/iam/v2/teams/spaces%20are%20borked
so as long as users are properly encoding their URIs in the API and the UI (which the UI now does correctly for both cases) things will work.

this is not a special case for our API in particular; however, we avoid this edge case in v2 by not allowing IDs with special characters.

also confirmed that migrated policies with IDs that contain special characters continue working on force upgrade.

tl;dr the UI previously didn't handle URI encoding properly for special case character links but does now for both the UI links and API requests. the encoding needed is simply standard URI encoding and isn't a special case for our tech. i don't think we need to mention anything to users (or even support at this point).


so closing...
